cni/network/network.go

```diff
@@ -487,20 +419,40 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		}
 	}
 
+	switch nwCfg.Ipam.Type {
+	case network.AzureCNS:
+		plugin.ipamInvoker, err = NewCNSInvoker(k8sPodName, k8sNamespace)
+		if err != nil {
+			log.Printf("[cni-net] Creating network %v.", networkId)
+			return err
+		}
+	default:
+		plugin.ipamInvoker = NewAzureIpamInvoker(plugin, &nwInfo)
+	}
+
+	options := make(map[string]interface{})
+
 	if nwInfoErr != nil {
 		// Network does not exist.
 		log.Printf("[cni-net] Creating network %v.", networkId)
 
 		if !nwCfg.MultiTenancy {
-			result, resultV6, err = plugin.invokeIpamAdd(*nwCfg, nwInfo, true)
+
+			options[optKeyCleanupDependency] = optValPool
+			result, resultV6, err = plugin.ipamInvoker.Add(args, nwCfg, &subnetPrefix, options)
 			if err != nil {
 				return err
 			}
 
 			defer func() {
 				if err != nil {
-					plugin.invokeIpamDel(result, nwCfg.Ipam.Type, *nwCfg, true)
-					plugin.invokeIpamDel(resultV6, ipamV6, *nwCfg, true)
+					options[optKeyCleanupDependency] = optValPool
+					if result != nil && result.IPs != nil && len(result.IPs) > 0 {
```

Comment:
no need to check if `result.IPs` is nil, `len(result.IPs)` is safe to use on nil slices